src/rhs_function.c

```diff
@@ -0,0 +1,498 @@
+    PetscReal dumr  = pow(hr[n], 0.5);
+    PetscReal cl    = pow(GRAVITY * hl[n], 0.5);
+    PetscReal cr    = pow(GRAVITY * hr[n], 0.5);
+    PetscReal hhat  = duml * dumr;
+    PetscReal uhat  = (duml * ul[n] + dumr * ur[n]) / (duml + dumr);
+    PetscReal vhat  = (duml * vl[n] + dumr * vr[n]) / (duml + dumr);
+    PetscReal chat  = pow(0.5 * GRAVITY * (hl[n] + hr[n]), 0.5);
+    PetscReal uperp = uhat * cn[n] + vhat * sn[n];
+
+    PetscReal dh     = hr[n] - hl[n];
+    PetscReal du     = ur[n] - ul[n];
+    PetscReal dv     = vr[n] - vl[n];
+    PetscReal dupar  = -du * sn[n] + dv * cn[n];
+    PetscReal duperp = du * cn[n] + dv * sn[n];
+
+    PetscReal dW[3];
+    dW[0] = 0.5 * (dh - hhat * duperp / chat);
+    dW[1] = hhat * dupar;
+    dW[2] = 0.5 * (dh + hhat * duperp / chat);
+
+    PetscReal uperpl = ul[n] * cn[n] + vl[n] * sn[n];
+    PetscReal uperpr = ur[n] * cn[n] + vr[n] * sn[n];
+    PetscReal al1    = uperpl - cl;
+    PetscReal al3    = uperpl + cl;
+    PetscReal ar1    = uperpr - cr;
+    PetscReal ar3    = uperpr + cr;
+
+    PetscReal R[3][3];
+    R[0][0] = 1.0;
+    R[0][1] = 0.0;
+    R[0][2] = 1.0;
+    R[1][0] = uhat - chat * cn[n];
+    R[1][1] = -sn[n];
+    R[1][2] = uhat + chat * cn[n];
+    R[2][0] = vhat - chat * sn[n];
+    R[2][1] = cn[n];
+    R[2][2] = vhat + chat * sn[n];
+
+    PetscReal da1 = fmax(0.0, 2.0 * (ar1 - al1));
+    PetscReal da3 = fmax(0.0, 2.0 * (ar3 - al3));
+    PetscReal a1  = fabs(uperp - chat);
+    PetscReal a2  = fabs(uperp);
+    PetscReal a3  = fabs(uperp + chat);
+
+    // Critical flow fix
+    if (a1 < da1) {
+      a1 = 0.5 * (a1 * a1 / da1 + da1);
+    }
+    if (a3 < da3) {
+      a3 = 0.5 * (a3 * a3 / da3 + da3);
+    }
+
+    // Compute interface flux
+    PetscReal A[3][3];
+    for (PetscInt i = 0; i < 3; i++) {
+      for (PetscInt j = 0; j < 3; j++) {
+        A[i][j] = 0.0;
+      }
+    }
+    A[0][0] = a1;
+    A[1][1] = a2;
+    A[2][2] = a3;
+
+    PetscReal FL[3], FR[3];
+    FL[0] = uperpl * hl[n];
+    FL[1] = ul[n] * uperpl * hl[n] + 0.5 * GRAVITY * hl[n] * hl[n] * cn[n];
+    FL[2] = vl[n] * uperpl * hl[n] + 0.5 * GRAVITY * hl[n] * hl[n] * sn[n];
+
+    FR[0] = uperpr * hr[n];
+    FR[1] = ur[n] * uperpr * hr[n] + 0.5 * GRAVITY * hr[n] * hr[n] * cn[n];
+    FR[2] = vr[n] * uperpr * hr[n] + 0.5 * GRAVITY * hr[n] * hr[n] * sn[n];
+
+    // fij = 0.5*(FL + FR - matmul(R,matmul(A,dW))
+    fij[n][0] = 0.5 * (FL[0] + FR[0] - R[0][0] * A[0][0] * dW[0] - R[0][1] * A[1][1] * dW[1] - R[0][2] * A[2][2] * dW[2]);
+    fij[n][1] = 0.5 * (FL[1] + FR[1] - R[1][0] * A[0][0] * dW[0] - R[1][1] * A[1][1] * dW[1] - R[1][2] * A[2][2] * dW[2]);
+    fij[n][2] = 0.5 * (FL[2] + FR[2] - R[2][0] * A[0][0] * dW[0] - R[2][1] * A[1][1] * dW[1] - R[2][2] * A[2][2] * dW[2]);
+
+    amax[n] = chat + fabs(uperp);
+  }
+
+  PetscFunctionReturn(0);
+}
+
+// computes RHS on internal edges
+static PetscErrorCode RHSFunctionForInternalEdges(RDy rdy, Vec F, PetscReal *amax_value) {
```

Comment:
Add the description https://github.com/RDycore/toy-problems/blob/main/swe_roe/ex2b.c#L1550

Reply:
Do we want to use Doxygen markup for internal RDycore functions, or is it okay just to document the parameters? I ask mostly because we're not taking a library approach with this project (outside of the stuff in the top-level `rdycore.h` file).